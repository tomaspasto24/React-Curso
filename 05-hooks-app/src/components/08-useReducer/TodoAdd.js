import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAddTodo } ) => {
    
    const [ { description }, handleInputChange, reset ] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 2 ){
            return;
        };

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input 
                    type='text'
                    name='description'
                    placeholder='Agregar tarea...'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />

                <div className = "d-grid gap-2">
                    <button 
                        type='submit'
                        className='btn btn-outline-primary mt-1'
                    >
                        Agregar
                    </button>
                </div>

            </form>  
        </>
    )
}
