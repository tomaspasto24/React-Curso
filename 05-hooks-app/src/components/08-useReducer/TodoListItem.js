import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ( { todo, index, handleDelete, handleToggle } ) => {
    return (
        <li
                key={todo.id}
                className='list-group-item'
            >

                <p
                    onClick={ () => handleToggle(todo.id) } 
                    className={ `${todo.done && 'complete'}` }
                >
                    { index+1 }. { todo.desc }
                </p>

                <button
                    onClick={ () => handleDelete(todo.id) } 
                    className='btn btn-danger'
                >
                    Borrar
                </button>

        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}