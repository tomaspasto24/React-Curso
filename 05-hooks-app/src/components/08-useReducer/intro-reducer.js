
const initialState = [{
    id:1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'agregar'){
        state = [ ...state, action.payload ]
    }

    return state;
};


let todos = todoReducer();

const newTodo = [{
    id:2,
    todo: 'salir a correr',
    done: false
}]

const action = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer(todos, action);

console.log(todos);