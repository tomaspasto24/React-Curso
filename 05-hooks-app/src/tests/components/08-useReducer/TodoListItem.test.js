import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';
import '@testing-library/jest-dom';

describe('Pruebas en TodoListItem.', () => {

        const handleDelete = jest.fn();
        const handleToggle = jest.fn();
        const todo = demoTodos[0];
    
        const wrapper = shallow( <TodoListItem 
                                        todo={ todo } 
                                        index={1}
                                        handleDelete={ handleDelete }
                                        handleToggle={ handleToggle }
                                 />);
        
    test('Debe mostrarse correctamente.', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de llamar a la función borrar.', () => {
        wrapper.find('button').simulate('click');

        expect( handleDelete ).toHaveBeenCalledWith(todo.id);
    });

    test('Debe de llamar a la función toggle.', () => {
        wrapper.find('p').simulate('click');

        expect( handleToggle ).toHaveBeenCalledWith(todo.id)
    });

    test('Debe de mostrar el texto correctamente.', () => {
        const textoP = wrapper.find('p').text();

        expect(textoP).toBe(`2. ${todo.desc}`);
    });

    test('Debe tener la clase complete si el done es true.', () => {
        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow( <TodoListItem 
            todo={ todo } 
            index={1}
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
            />);

        expect( wrapper.find('p').hasClass('complete') ).toBe(true);
    });

});