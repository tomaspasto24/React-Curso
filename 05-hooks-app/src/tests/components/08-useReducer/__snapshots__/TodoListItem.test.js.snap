// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoListItem. Debe mostrarse correctamente. 1`] = `
<li
  className="list-group-item"
  key="2"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2
    . 
    Aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
